lib/unifex/code_generator/base_types/list.ex

```diff
@@ -27,66 +59,112 @@ defmodule Unifex.CodeGenerator.BaseTypes.List do
+  end
+
+  defmodule CNode do
+    @moduledoc false
+    use Unifex.CodeGenerator.BaseType
+    alias Unifex.CodeGenerator.BaseType
 
     @impl BaseType
-    def generate_initialization(name, _ctx) do
-      ~g<#{name} = NULL;>
+    def generate_arg_serialize(name, ctx) do
+      ~g"""
+      ({
+        ei_x_encode_list_header(out_buff, #{name}_length);
+        for(unsigned int i = 0; i < #{name}_length; i++) {
+          #{BaseType.generate_arg_serialize(ctx.subtype, :"#{name}[i]", ctx.generator)}
+        }
+        ei_x_encode_empty_list(out_buff);
+      });
+      """
     end
 
     @impl BaseType
-    def generate_destruction(name, ctx) do
+    def generate_arg_parse(arg, var_name, ctx) do
+      elem_name = :"#{var_name}[i]"
+      len_var_name = "#{var_name}_length"
+      native_type = BaseType.generate_native_type(ctx.subtype, ctx.generator)
+      %{subtype: subtype, postproc_fun: postproc_fun, generator: generator} = ctx
+
       ~g"""
-      if(#{name} != NULL) {
-        for(unsigned int i = 0; i < #{name}_length; i++) {
-          #{BaseType.generate_destruction(ctx.subtype, :"#{name}[i]", ctx.generator)}
+      ({
+        int res = 1;
+        int type;
+        int size;
+
+        ei_get_type(#{arg}->buff, #{arg}->index, &type, &size);
+        #{len_var_name} = (unsigned int) size;
+
+        UnifexCNodeInBuff *unifex_buff;
+        unifex_buff = malloc(sizeof(UnifexCNodeInBuff));
+        if(type == ERL_STRING_EXT) {
+          char *p = malloc(sizeof(char) * #{len_var_name});
+          res = ei_decode_string(#{arg}->buff, #{arg}->index, p);
+          unsigned char *unsigned_p = (unsigned char *)p;
+
+          ei_x_buff buff;
+          ei_x_new_with_version(&buff);
+          ei_x_encode_list_header(&buff, #{len_var_name});
+          for(unsigned int i = 0; i < #{len_var_name}; i++) {
+            ei_x_encode_ulong(&buff, (unsigned long)unsigned_p[i]);
+          }
+          ei_x_encode_empty_list(&buff);
+
+          int version;
+          int index = 0;
+          ei_decode_version(buff.buff, &index, &version);
+          res = ei_get_type(buff.buff, &index, &type, &size);
+          unifex_buff->buff = buff.buff;
+          unifex_buff->index = &index;
+        } else {
+          unifex_buff->buff = #{arg}->buff;
+          unifex_buff->index = #{arg}->index;
+        }
+
+        res = ei_decode_list_header(unifex_buff->buff, unifex_buff->index, &size);
+        #{len_var_name} = (unsigned int) size;
+        #{var_name} = malloc(sizeof(#{native_type}) * #{len_var_name});
+
+        for(unsigned int i = 0; i < #{len_var_name}; i++) {
+          #{BaseType.generate_initialization(subtype, elem_name, generator)}
         }
-        enif_free(#{name});
+
+        for(unsigned int i = 0; i < #{len_var_name}; i++) {
+          #{
+        BaseType.generate_arg_parse(subtype, elem_name, "unifex_buff", postproc_fun, generator)
       }
+        }
+        res = ei_skip_term(unifex_buff->buff, unifex_buff->index);
```

Comment:
This one line can seem a little tricky. It's for omitting empty list that is encoded at list's end. Added test for it.

Reply:
you can use `ei_decode_empty_list`